File: app/src/main/java/jp/mimac/urlsaver/domain/UrlRules.kt
package jp.mimac.urlsaver.domain

import android.content.Intent
import java.security.MessageDigest
import java.net.URI
import java.util.Locale

object UrlRules {
    const val MAX_INPUT_TEXT_BYTES: Int = 256 * 1024
    const val MAX_EXTRACTED_URLS_PER_INPUT: Int = 50
    const val MAX_BATCH_SAVE_URLS_PER_INTAKE: Int = 50
    const val TEXT_CARD_HOST: String = "text.rinbam.local"

    data class ExtractedUrlBatch(
        val urls: List<String>,
        val truncatedToMaxUrls: Boolean,
        val sawOversizedInput: Boolean,
    )

    fun extractFromIntent(intent: Intent): ShareExtractionResult {
        val sources = extractCandidateSources(intent)
        var hasAnyCandidateString = false
        var sawOversizedInput = false

        for (sourceCandidates in sources) {
            if (sourceCandidates.isEmpty()) continue
            hasAnyCandidateString = true

            val validInSource = findFirstValidUrl(sourceCandidates)
            when (validInSource) {
                is ShareExtractionResult.Found -> return validInSource
                ShareExtractionResult.InputTooLarge -> sawOversizedInput = true
                ShareExtractionResult.InvalidUrl,
                ShareExtractionResult.NoUrlFound,
                null -> Unit
            }
        }

        return when {
            sawOversizedInput -> ShareExtractionResult.InputTooLarge
            hasAnyCandidateString -> ShareExtractionResult.InvalidUrl
            else -> ShareExtractionResult.NoUrlFound
        }
    }

    fun countValidUrlsInIntent(intent: Intent): Int {
        return extractAllFromIntent(intent).urls.size
    }

    fun extractAllFromIntent(intent: Intent): ExtractedUrlBatch {
        val normalizedUrls = linkedSetOf<String>()
        var truncatedToMaxUrls = false
        var sawOversizedInput = false

        outer@ for (sourceCandidates in extractCandidateSources(intent)) {
            for (candidateText in sourceCandidates) {
                if (!isWithinInputTextByteLimit(candidateText)) {
                    sawOversizedInput = true
                    continue
                }
                for (candidateUrl in extractUrlCandidates(candidateText)) {
                    val normalized = normalize(candidateUrl) ?: continue
                    val inserted = normalizedUrls.add(normalized)
                    if (!inserted) continue
                    if (normalizedUrls.size > MAX_EXTRACTED_URLS_PER_INPUT) {
                        normalizedUrls.remove(normalized)
                        truncatedToMaxUrls = true
                        break@outer
                    }
                }
            }
        }
        return ExtractedUrlBatch(
            urls = normalizedUrls.toList().take(MAX_BATCH_SAVE_URLS_PER_INTAKE),
            truncatedToMaxUrls = truncatedToMaxUrls || normalizedUrls.size > MAX_BATCH_SAVE_URLS_PER_INTAKE,
            sawOversizedInput = sawOversizedInput,
        )
    }

    fun extractForManualInput(input: String): ShareExtractionResult {
        val trimmed = input.trim()
        if (trimmed.isEmpty()) return ShareExtractionResult.NoUrlFound
        if (!isWithinInputTextByteLimit(trimmed)) return ShareExtractionResult.InputTooLarge
        val valid = findFirstValidUrl(listOf(trimmed))
        return when (valid) {
            is ShareExtractionResult.Found -> valid
            ShareExtractionResult.InputTooLarge -> ShareExtractionResult.InputTooLarge
            ShareExtractionResult.InvalidUrl,
            ShareExtractionResult.NoUrlFound,
            null -> {
                if (extractUrlCandidates(trimmed).isEmpty()) {
                    ShareExtractionResult.NoUrlFound
                } else {
                    ShareExtractionResult.InvalidUrl
                }
            }
        }
    }

    fun extractTextFallbackFromIntent(intent: Intent): String? {
        return extractCandidateSources(intent)
            .asSequence()
            .flatten()
            .mapNotNull(::normalizeTextCardBody)
            .firstOrNull()
    }

    fun extractMemoWithoutUrlsFromIntent(intent: Intent): String? {
        return extractCandidateSources(intent)
            .asSequence()
            .flatten()
            .mapNotNull(::extractMemoWithoutUrls)
            .firstOrNull()
    }

    fun extractMemoWithoutUrls(input: String): String? {
        val normalized = normalizeTextCardBody(input) ?: return null
        val ranges = URL_CANDIDATE_REGEX
            .findAll(normalized)
            .filter { normalize(it.value) != null }
            .map { it.range }
            .toList()
        if (ranges.isEmpty()) return null

        val withoutUrls = buildString {
            var index = 0
            for (range in ranges) {
                if (index < range.first) {
                    append(normalized.substring(index, range.first))
                }
                append('\n')
                index = range.last + 1
            }
            if (index < normalized.length) {
                append(normalized.substring(index))
            }
        }
        val memo = withoutUrls
            .lineSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .joinToString("\n")
            .trim()
        if (memo.isBlank()) return null
        return memo.take(2000)
    }

    fun normalizeTextCardBody(input: String): String? {
        val normalized = input
            .replace("\r\n", "\n")
            .replace('\r', '\n')
            .trim()
        if (normalized.isEmpty()) return null
        if (!isWithinInputTextByteLimit(normalized)) return null
        return normalized
    }

    fun parseTextCard(input: String): ParsedUrl? {
        val body = normalizeTextCardBody(input) ?: return null
        val hash = sha256(body)
        val normalized = "https://$TEXT_CARD_HOST/note/$hash"
        return ParsedUrl(
            originalUrl = body,
            normalizedUrl = normalized,
            displayUrl = "テキスト",
            openUrl = normalized,
            normalizedHost = TEXT_CARD_HOST,
            rawSourceHost = TEXT_CARD_HOST,
            serviceType = ServiceType.WEB,
            contentContext = ContentContext.POST,
        )
    }

    fun isTextCardHost(host: String?): Boolean {
        return host?.equals(TEXT_CARD_HOST, ignoreCase = true) == true
    }

    fun textCardTitle(body: String): String {
        val firstLine = body
            .lineSequence()
            .map { it.trim() }
            .firstOrNull { it.isNotBlank() }
            ?: body.trim()
        return firstLine
            .replace(Regex("\\s+"), " ")
            .take(120)
            .ifBlank { "テキスト" }
    }

    fun parseUrl(original: String): ParsedUrl? {
        val normalized = normalize(original) ?: return null
        val uri = URI(normalized)
        val host = uri.host.orEmpty()
        val service = classifyService(host)
        val contentContext = classifyContent(service, uri)
        val display = toDisplayUrl(normalized, service)
        return ParsedUrl(
            originalUrl = original,
            normalizedUrl = normalized,
            displayUrl = display,
            openUrl = normalized,
            normalizedHost = host,
            rawSourceHost = host,
            serviceType = service,
            contentContext = contentContext,
        )
    }

    fun normalize(raw: String): String? {
        val value = raw.trim()
        val uri = try {
            URI(value)
        } catch (_: Exception) {
            return null
        }

        val scheme = uri.scheme?.lowercase(Locale.ROOT) ?: return null
        if (!uri.isAbsolute) return null

        val host = uri.host?.lowercase(Locale.ROOT)?.takeIf { it.isNotBlank() } ?: return null
        if (scheme != "https" && !(scheme == "http" && isLoopbackHost(host))) {
            return null
        }
        val userInfo = uri.userInfo

        val port = when {
            scheme == "http" && uri.port == 80 -> -1
            scheme == "https" && uri.port == 443 -> -1
            else -> uri.port
        }

        var path = uri.rawPath.orEmpty().ifEmpty { "/" }
        if (path != "/" && path.endsWith("/")) {
            path = path.trimEnd('/')
            if (path.isEmpty()) path = "/"
        }

        return try {
            URI(
                scheme,
                userInfo,
                host,
                port,
                path,
                uri.rawQuery,
                null,
            ).toASCIIString()
        } catch (_: Exception) {
            null
        }
    }

    fun toDisplayUrl(normalizedUrl: String, service: ServiceType): String {
        val uri = URI(normalizedUrl)
        val host = uri.host.orEmpty()
        val path = uri.rawPath.orEmpty().ifEmpty { "/" }
        val query = when (service) {
            ServiceType.YOUTUBE -> extractYoutubeQuery(uri.rawQuery)
            else -> null
        }
        return buildString {
            append(host)
            append(path)
            if (!query.isNullOrEmpty()) {
                append('?')
                append(query)
            }
        }
    }

    fun toLegacyHttpTwin(normalizedUrl: String): String? {
        val uri = runCatching { URI(normalizedUrl) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase(Locale.ROOT) ?: return null
        if (scheme != "https") return null
        val host = uri.host?.takeIf { it.isNotBlank() } ?: return null
        return runCatching {
            URI(
                "http",
                uri.userInfo,
                host,
                uri.port,
                uri.rawPath,
                uri.rawQuery,
                null,
            ).toASCIIString()
        }.getOrNull()
    }

    fun metadataFetchUrls(openUrl: String, @Suppress("UNUSED_PARAMETER") serviceType: ServiceType): List<String> {
        return listOf(openUrl)
    }

    fun extractXStatusId(url: String): String? {
        val uri = runCatching { URI(url) }.getOrNull() ?: return null
        val host = uri.host?.lowercase(Locale.ROOT) ?: return null
        if (!isXStatusProviderHost(host)) return null
        return extractXStatusPath(uri)?.statusId
    }

    private fun extractYoutubeQuery(query: String?): String? {
        if (query.isNullOrEmpty()) return null
        val keep = query.split('&').firstOrNull { it.startsWith("v=") }
        return keep?.takeIf { it.length > 2 }
    }

    private fun findFirstValidUrl(sourceCandidates: List<String>): ShareExtractionResult? {
        var sawOversizedInput = false
        var scannedCandidateUrls = 0
        for (candidateText in sourceCandidates) {
            if (!isWithinInputTextByteLimit(candidateText)) {
                sawOversizedInput = true
                continue
            }
            val extracted = extractUrlCandidates(candidateText)
            for (candidateUrl in extracted) {
                scannedCandidateUrls += 1
                if (scannedCandidateUrls > MAX_EXTRACTED_URLS_PER_INPUT) {
                    return null
                }
                if (normalize(candidateUrl) != null) {
                    return ShareExtractionResult.Found(candidateUrl)
                }
            }
        }
        return if (sawOversizedInput) ShareExtractionResult.InputTooLarge else null
    }

    private fun extractUrlCandidates(text: String): List<String> {
        return URL_CANDIDATE_REGEX.findAll(text).map { it.value }.toList()
    }

    private fun isWithinInputTextByteLimit(text: String): Boolean {
        return text.toByteArray(Charsets.UTF_8).size <= MAX_INPUT_TEXT_BYTES
    }

    private fun extractCandidateSources(intent: Intent): List<List<String>> {
        val extraCandidates = mutableListOf<String>().apply {
            intent.getStringExtra(Intent.EXTRA_TEXT)?.let(::add)
            intent.getCharSequenceExtra(Intent.EXTRA_TEXT)?.toString()?.let(::add)
            intent.getStringArrayListExtra(Intent.EXTRA_TEXT)?.forEach(::add)
            intent.getCharSequenceArrayListExtra(Intent.EXTRA_TEXT)
                ?.forEach { add(it.toString()) }
            intent.getStringExtra(Intent.EXTRA_HTML_TEXT)?.let(::add)
            intent.getStringExtra(Intent.EXTRA_SUBJECT)?.let(::add)
            intent.getCharSequenceExtra(Intent.EXTRA_SUBJECT)?.toString()?.let(::add)
        }
        val clipCandidates = mutableListOf<String>().apply {
            val clip = intent.clipData ?: return@apply
            for (i in 0 until clip.itemCount) {
                val item = clip.getItemAt(i)
                item.text?.toString()?.let(::add)
                item.uri?.toString()?.let(::add)
            }
        }
        @Suppress("DEPRECATION")
        val streamCandidates = mutableListOf<String>().apply {
            intent.getParcelableExtra<android.net.Uri>(Intent.EXTRA_STREAM)
                ?.let { add(it.toString()) }
            intent.getParcelableArrayListExtra<android.net.Uri>(Intent.EXTRA_STREAM)
                ?.forEach { add(it.toString()) }
        }
        val dataCandidates = mutableListOf<String>().apply {
            intent.dataString?.let(::add)
        }
        return listOf(extraCandidates, clipCandidates, streamCandidates, dataCandidates)
    }

    private fun isLoopbackHost(host: String): Boolean {
        return host == "127.0.0.1" || host == "localhost" || host == "::1"
    }

    private fun sha256(value: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(value.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { byte -> "%02x".format(byte) }
    }

    private fun isXHost(host: String): Boolean {
        val lowered = host.lowercase(Locale.ROOT)
        return lowered == "x.com" || lowered.endsWith("twitter.com")
    }

    private fun isXStatusProviderHost(host: String): Boolean {
        val lowered = host.lowercase(Locale.ROOT)
        return lowered == "x.com" ||
            lowered.endsWith("twitter.com")
    }

    private fun classifyService(host: String): ServiceType {
        val lowered = host.lowercase(Locale.ROOT)
        return when {
            lowered.endsWith("youtube.com") || lowered == "youtu.be" -> ServiceType.YOUTUBE
            lowered.endsWith("tiktok.com") -> ServiceType.TIKTOK
            isXHost(host) -> ServiceType.X
            lowered.endsWith("instagram.com") -> ServiceType.INSTAGRAM
            else -> ServiceType.WEB
        }
    }

    private fun extractXStatusPath(uri: URI): XStatusPath? {
        val segments = uri.path.orEmpty().split('/').filter { it.isNotBlank() }
        if (segments.size >= 3 && segments[1].equals("status", ignoreCase = true)) {
            val screenName = segments[0].takeIf { it.isNotBlank() }
            val statusId = segments[2].takeIf { it.isNotBlank() } ?: return null
            return XStatusPath(screenName = screenName, statusId = statusId)
        }
        if (segments.size >= 4 &&
            segments[0].equals("i", ignoreCase = true) &&
            segments[1].equals("web", ignoreCase = true) &&
            segments[2].equals("status", ignoreCase = true)
        ) {
            val statusId = segments[3].takeIf { it.isNotBlank() } ?: return null
            return XStatusPath(screenName = null, statusId = statusId)
        }
        return null
    }

    private fun classifyContent(service: ServiceType, uri: URI): ContentContext {
        val path = uri.path.orEmpty().lowercase(Locale.ROOT)
        return when (service) {
            ServiceType.YOUTUBE -> when {
                path.startsWith("/shorts/") -> ContentContext.SHORTS
                path.startsWith("/live/") -> ContentContext.LIVE
                path.startsWith("/watch") || uri.query?.contains("v=") == true -> ContentContext.VIDEO
                else -> ContentContext.STANDARD
            }
            ServiceType.TIKTOK -> when {
                path.contains("/video/") -> ContentContext.VIDEO
                path.contains("/music/") -> ContentContext.MUSIC
                path.contains("/tag/") -> ContentContext.HASHTAG
                else -> ContentContext.STANDARD
            }
            ServiceType.X -> when {
                path.contains("/status/") -> ContentContext.POST
                else -> ContentContext.STANDARD
            }
            ServiceType.INSTAGRAM -> when {
                path.startsWith("/reel/") -> ContentContext.REEL
                path.startsWith("/p/") -> ContentContext.POST
                path.startsWith("/@") -> ContentContext.PROFILE
                else -> ContentContext.STANDARD
            }
            ServiceType.WEB -> ContentContext.STANDARD
            ServiceType.ALL -> ContentContext.STANDARD
        }
    }

    fun effectiveTitle(
        userTitle: String?,
        fetchedTitle: String?,
        serviceType: ServiceType,
        normalizedHost: String,
    ): String {
        val titleService = if (serviceType == ServiceType.TIKTOK) ServiceType.WEB else serviceType
        return when {
            !userTitle.isNullOrBlank() -> userTitle
            !fetchedTitle.isNullOrBlank() -> fetchedTitle
            titleService != ServiceType.WEB -> "${titleService.displayName}のリンク"
            normalizedHost.isNotBlank() -> normalizedHost
            else -> "保存したリンク"
        }
    }

    fun normalizeUserTitle(input: String?): String? {
        val trimmed = input?.trim() ?: return null
        if (trimmed.isBlank()) return null
        if (trimmed.length > 120) return null
        return trimmed
    }

    fun normalizeMemo(input: String?): String {
        val trimmed = input?.trim().orEmpty()
        if (trimmed.isBlank()) return ""
        return trimmed
    }

    fun isMemoLengthValid(input: String): Boolean = input.trim().length <= 2000
    fun isTitleLengthValid(input: String): Boolean = input.trim().length <= 120

    private data class XStatusPath(
        val screenName: String?,
        val statusId: String,
    )

    private val URL_CANDIDATE_REGEX = Regex("https?://[^\\s<>()\\[\\]\"']+", RegexOption.IGNORE_CASE)
}

File: ios/URLSaverShared/Domain/URLRules.swift
import Foundation
import CryptoKit

enum URLRules {
    static let maxInputTextBytes = 256 * 1024
    static let maxExtractedURLsPerInput = 50
    static let maxBatchSaveURLsPerIntake = 50
    static let textCardHost = "text.rinbam.local"

    struct ExtractedURLBatch: Equatable, Sendable {
        let urls: [String]
        let truncatedToMaxURLs: Bool
        let sawOversizedInput: Bool
    }

    static func extractFromCandidateGroups(_ candidateGroups: ShareCandidateGroups) -> ShareExtractionResult {
        var hasAnyCandidateString = false
        var sawOversizedInput = false

        for sourceCandidates in candidateGroups.orderedGroups {
            if sourceCandidates.isEmpty { continue }
            hasAnyCandidateString = true

            switch findFirstValidURL(in: sourceCandidates) {
            case .found(let valid):
                return .found(valid)
            case .inputTooLarge:
                sawOversizedInput = true
            case .noURLFound, .invalidURL:
                break
            case nil:
                break
            }
        }

        if sawOversizedInput { return .inputTooLarge }
        return hasAnyCandidateString ? .invalidURL : .noURLFound
    }

    static func extractAllFromCandidateGroups(_ candidateGroups: ShareCandidateGroups) -> ExtractedURLBatch {
        var normalizedURLs = OrderedSet<String>()
        var truncatedToMaxURLs = false
        var sawOversizedInput = false

        for sourceCandidates in candidateGroups.orderedGroups {
            for candidateText in sourceCandidates {
                guard isWithinInputTextByteLimit(candidateText) else {
                    sawOversizedInput = true
                    continue
                }
                for candidateURL in extractURLCandidates(from: candidateText) {
                    guard let normalized = normalize(candidateURL) else { continue }
                    guard normalizedURLs.append(normalized) else { continue }
                    if normalizedURLs.elements.count > maxExtractedURLsPerInput {
                        normalizedURLs.removeLast()
                        truncatedToMaxURLs = true
                        return ExtractedURLBatch(
                            urls: Array(normalizedURLs.elements.prefix(maxBatchSaveURLsPerIntake)),
                            truncatedToMaxURLs: true,
                            sawOversizedInput: sawOversizedInput
                        )
                    }
                }
            }
        }
        return ExtractedURLBatch(
            urls: Array(normalizedURLs.elements.prefix(maxBatchSaveURLsPerIntake)),
            truncatedToMaxURLs: truncatedToMaxURLs || normalizedURLs.elements.count > maxBatchSaveURLsPerIntake,
            sawOversizedInput: sawOversizedInput
        )
    }

    static func countValidURLs(in candidateGroups: ShareCandidateGroups) -> Int {
        extractAllFromCandidateGroups(candidateGroups).urls.count
    }

    static func extractForManualInput(_ input: String) -> ShareExtractionResult {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return .noURLFound
        }

        guard isWithinInputTextByteLimit(trimmed) else {
            return .inputTooLarge
        }

        switch findFirstValidURL(in: [trimmed]) {
        case .found(let valid):
            return .found(valid)
        case .inputTooLarge:
            return .inputTooLarge
        case .noURLFound, .invalidURL:
            return .invalidURL
        case nil:
            return extractURLCandidates(from: trimmed).isEmpty ? .noURLFound : .invalidURL
        }
    }

    static func extractTextFallback(from candidateGroups: ShareCandidateGroups) -> String? {
        for group in candidateGroups.orderedGroups {
            for candidate in group {
                if let text = normalizeTextCardBody(candidate) {
                    return text
                }
            }
        }
        return nil
    }

    static func extractMemoWithoutURLs(from candidateGroups: ShareCandidateGroups) -> String? {
        for group in candidateGroups.orderedGroups {
            for candidate in group {
                if let memo = extractMemoWithoutURLs(candidate) {
                    return memo
                }
            }
        }
        return nil
    }

    static func extractMemoWithoutURLs(_ input: String) -> String? {
        guard let normalized = normalizeTextCardBody(input) else { return nil }
        let nsText = normalized as NSString
        let validRanges = extractURLCandidateRanges(from: normalized)
            .filter { range in
                normalize(nsText.substring(with: range)) != nil
            }
        guard !validRanges.isEmpty else { return nil }

        var pieces: [String] = []
        var index = 0
        for range in validRanges {
            if index < range.location {
                pieces.append(nsText.substring(with: NSRange(location: index, length: range.location - index)))
            }
            pieces.append("\n")
            index = range.location + range.length
        }
        if index < nsText.length {
            pieces.append(nsText.substring(from: index))
        }

        let memo = pieces
            .joined()
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !memo.isEmpty else { return nil }
        return String(memo.prefix(2_000))
    }

    static func normalizeTextCardBody(_ input: String) -> String? {
        let normalized = input
            .replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty, isWithinInputTextByteLimit(normalized) else {
            return nil
        }
        return normalized
    }

    static func parseTextCard(_ input: String) -> ParsedURL? {
        guard let body = normalizeTextCardBody(input) else { return nil }
        let hash = sha256(body)
        let normalized = "https://\(textCardHost)/note/\(hash)"
        return ParsedURL(
            originalURL: body,
            normalizedURL: normalized,
            displayURL: "テキスト",
            openURL: normalized,
            normalizedHost: textCardHost,
            rawSourceHost: textCardHost,
            serviceType: .web,
            contentContext: .post
        )
    }

    static func isTextCardHost(_ host: String?) -> Bool {
        host?.caseInsensitiveCompare(textCardHost) == .orderedSame
    }

    static func textCardTitle(_ body: String) -> String {
        let firstLine = body
            .split(whereSeparator: \.isNewline)
            .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
            .first(where: { !$0.isEmpty }) ?? body.trimmingCharacters(in: .whitespacesAndNewlines)
        let collapsed = firstLine
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
        let title = String(collapsed.prefix(120))
        return title.isEmpty ? "テキスト" : title
    }

    static func parseURL(_ original: String) -> ParsedURL? {
        guard let normalized = normalize(original),
              let components = URLComponents(string: normalized),
              let host = components.host,
              !host.isEmpty else {
            return nil
        }

        let service = classifyService(host: host)
        let contentContext = classifyContent(service: service, components: components)
        let display = toDisplayURL(normalizedURL: normalized, service: service)
        return ParsedURL(
            originalURL: original,
            normalizedURL: normalized,
            displayURL: display,
            openURL: normalized,
            normalizedHost: host,
            rawSourceHost: host,
            serviceType: service,
            contentContext: contentContext
        )
    }

    static func normalize(_ raw: String) -> String? {
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: value),
              let scheme = components.scheme?.lowercased(),
              let host = components.host?.lowercased(),
              !host.isEmpty else {
            return nil
        }

        let absolute = value.contains("://")
        guard absolute else { return nil }
        guard scheme == "https" || (scheme == "http" && isLoopbackHost(host)) else {
            return nil
        }
        guard components.user == nil, components.password == nil else { return nil }

        components.scheme = scheme
        components.host = host
        components.fragment = nil

        if (scheme == "https" && components.port == 443) || (scheme == "http" && components.port == 80) {
            components.port = nil
        }

        var path = components.percentEncodedPath.isEmpty ? "/" : components.percentEncodedPath
        if path != "/" && path.hasSuffix("/") {
            path = String(path.drop(while: { $0 == "/" }).dropLast())
            path = "/" + path
            if path == "/" {
                path = "/"
            }
        }
        components.percentEncodedPath = path

        return components.url?.absoluteString
    }

    static func toDisplayURL(normalizedURL: String, service: ServiceType) -> String {
        guard let components = URLComponents(string: normalizedURL),
              let host = components.host else {
            return normalizedURL
        }
        let path = components.percentEncodedPath.isEmpty ? "/" : components.percentEncodedPath
        let query: String?
        switch service {
        case .youtube:
            query = components.queryItems?
                .first(where: { $0.name == "v" && ($0.value?.isEmpty == false) })
                .flatMap { item in
                    guard let value = item.value else { return nil }
                    return "v=\(value)"
                }
        default:
            query = nil
        }

        if let query {
            return "\(host)\(path)?\(query)"
        }
        return "\(host)\(path)"
    }

    static func effectiveTitle(
        userTitle: String?,
        fetchedTitle: String?,
        serviceType: ServiceType,
        normalizedHost: String
    ) -> String {
        if let userTitle = userTitle, !userTitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return userTitle
        }
        if let fetchedTitle = fetchedTitle, !fetchedTitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return fetchedTitle
        }
        if isTextCardHost(normalizedHost) {
            return "テキスト"
        }
        if serviceType != .web {
            return "\(serviceType.displayName)のリンク"
        }
        if !normalizedHost.isEmpty {
            return normalizedHost
        }
        return "保存したリンク"
    }

    static func normalizeUserTitle(_ input: String?) -> String? {
        guard let trimmed = input?.trimmingCharacters(in: .whitespacesAndNewlines),
              !trimmed.isEmpty,
              trimmed.count <= 120 else {
            return nil
        }
        return trimmed
    }

    static func normalizeMemo(_ input: String?) -> String {
        let trimmed = input?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? "" : trimmed
    }

    static func isMemoLengthValid(_ input: String) -> Bool {
        input.trimmingCharacters(in: .whitespacesAndNewlines).count <= 2_000
    }

    static func isTitleLengthValid(_ input: String) -> Bool {
        input.trimmingCharacters(in: .whitespacesAndNewlines).count <= 120
    }

    static func extractXStatusID(from url: String) -> String? {
        guard let components = URLComponents(string: url),
              let host = components.host?.lowercased(),
              host == "x.com" || host.hasSuffix("twitter.com") else {
            return nil
        }

        let segments = components.path.split(separator: "/").map(String.init)
        if segments.count >= 3, segments[1].lowercased() == "status" {
            return segments[2]
        }
        if segments.count >= 4,
           segments[0].lowercased() == "i",
           segments[1].lowercased() == "web",
           segments[2].lowercased() == "status" {
            return segments[3]
        }
        if segments.count >= 3,
           segments[0].lowercased() == "i",
           segments[1].lowercased() == "status" {
            return segments[2]
        }
        return nil
    }

    private static func findFirstValidURL(in sourceCandidates: [String]) -> ShareExtractionResult? {
        var sawOversizedInput = false
        var scannedCandidateURLs = 0
        for candidateText in sourceCandidates {
            guard isWithinInputTextByteLimit(candidateText) else {
                sawOversizedInput = true
                continue
            }
            for candidateURL in extractURLCandidates(from: candidateText) {
                scannedCandidateURLs += 1
                if scannedCandidateURLs > maxExtractedURLsPerInput {
                    return nil
                }
                if normalize(candidateURL) != nil {
                    return .found(candidateURL)
                }
            }
        }
        return sawOversizedInput ? .inputTooLarge : nil
    }

    private static func sha256(_ value: String) -> String {
        let digest = SHA256.hash(data: Data(value.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func extractURLCandidates(from text: String) -> [String] {
        let nsText = text as NSString
        return extractURLCandidateRanges(from: text).map {
            nsText.substring(with: $0)
        }
    }

    private static func extractURLCandidateRanges(from text: String) -> [NSRange] {
        let pattern = #"https?://[^\s<>()\[\]"']+"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return []
        }
        let nsText = text as NSString
        return regex.matches(in: text, range: NSRange(location: 0, length: nsText.length)).map(\.range)
    }

    private static func isLoopbackHost(_ host: String) -> Bool {
        host == "127.0.0.1" || host == "localhost" || host == "::1"
    }

    private static func isWithinInputTextByteLimit(_ text: String) -> Bool {
        text.lengthOfBytes(using: .utf8) <= maxInputTextBytes
    }

    private static func classifyService(host: String) -> ServiceType {
        let lowered = host.lowercased()
        switch lowered {
        case _ where lowered.hasSuffix("youtube.com") || lowered == "youtu.be":
            return .youtube
        case _ where lowered.hasSuffix("tiktok.com"):
            return .tiktok
        case _ where lowered == "x.com" || lowered.hasSuffix("twitter.com"):
            return .x
        case _ where lowered.hasSuffix("instagram.com"):
            return .instagram
        default:
            return .web
        }
    }

    private static func classifyContent(service: ServiceType, components: URLComponents) -> ContentContext {
        let path = components.path.lowercased()
        switch service {
        case .youtube:
            if path.hasPrefix("/shorts/") { return .shorts }
            if path.hasPrefix("/live/") { return .live }
            if path.hasPrefix("/watch") || components.queryItems?.contains(where: { $0.name == "v" }) == true {
                return .video
            }
            return .standard
        case .tiktok:
            if path.contains("/video/") { return .video }
            if path.contains("/music/") { return .music }
            if path.contains("/tag/") { return .hashtag }
            return .standard
        case .x:
            return path.contains("/status/") ? .post : .standard
        case .instagram:
            if path.hasPrefix("/reel/") { return .reel }
            if path.hasPrefix("/p/") { return .post }
            if path.hasPrefix("/@") { return .profile }
            return .standard
        case .web, .all:
            return .standard
        }
    }
}

private struct OrderedSet<Element: Hashable> {
    private var seen = Set<Element>()
    private(set) var elements: [Element] = []

    @discardableResult
    mutating func append(_ element: Element) -> Bool {
        guard seen.insert(element).inserted else { return false }
        elements.append(element)
        return true
    }

    mutating func removeLast() {
        guard let last = elements.popLast() else { return }
        seen.remove(last)
    }
}
